includes/admin/services/class-rop-gmb-service.php

```diff
@@ -0,0 +1,512 @@
+	 * @access public
+	 *
+	 * @param string $app_id The APP ID. Default empty.
+	 * @param string $secret The APP Secret. Default empty.
+	 *
+	 * @return null abstract method not used for this service specifically.
+	 */
+	public function get_api( $app_id = '', $secret = '' ) {
+		return;
+	}
+
+	/**
+	 * Abstract function, not in Use. Method to define the api.
+	 *
+	 * @since  8.5.9
+	 * @access public
+	 *
+	 * @param string $app_id The APP ID. Default empty.
+	 * @param string $secret The APP Secret. Default empty.
+	 *
+	 * @return mixed
+	 */
+	public function set_api( $app_id = '', $secret = '' ) {
+		return;
+	}
+
+	/**
+	 * Abstract function, not in Use. Method for authenticate the service.
+	 *
+	 * @codeCoverageIgnore
+	 *
+	 * @since  8.5.9
+	 * @access public
+	 * @return mixed
+	 */
+	public function maybe_authenticate() {
+		return;
+	}
+
+	/**
+	 * Abstract function, not in Use. Method to authenticate an user based on provided credentials.
+	 * Used in DB upgrade.
+	 *
+	 * @param array $args The arguments for service auth.
+	 *
+	 * @return bool
+	 */
+	public function authenticate( $args = array() ) {
+		return;
+	}
+
+	/**
+	 * This method will load and prepare the account data for Google My Business user.
+	 * Used in Rest Api.
+	 *
+	 * @since   8.5.9
+	 * @access  public
+	 *
+	 * @param   array $accounts_data Buffer accounts data.
+	 *
+	 * @return  bool
+	 */
+	public function add_account_with_app( $accounts_data ) {
+
+		if ( ! $this->is_set_not_empty( $accounts_data, array( 'id' ) ) ) {
+			$this->logger->alert_error( Rop_I18n::get_labels( 'errors.gmb_no_valid_accounts' ) );
+			return false;
+		}
+
+		$the_id       = unserialize( base64_decode( $accounts_data['id'] ) );
+		$accounts_array = unserialize( base64_decode( $accounts_data['pages'] ) );
+
+		$accounts = array();
+
+		for ( $i = 0; $i < sizeof( $accounts_array ); $i++ ) {
+
+			$account = $this->user_default;
+
+			$account_data = $accounts_array[ $i ];
+
+			$account['id'] = $account_data['id'];
+			$account['img'] = $account_data['img'];
+			$account['account'] = $account_data['account'];
+			$account['user'] = $account_data['user'];
+
+			if ( $i === 0 ) {
+				$account['active'] = true;
+			} else {
+				$account['active'] = false;
+			}
+
+			$accounts[] = $account;
+		}
+
+		// Prepare the data that will be saved as new account added.
+		$this->service = array(
+			'id'                 => $the_id,
+			'service'            => $this->service_name,
+			'credentials'        => array(
+				'created'                => $account_data['created'],
+				'expires_in'         => $account_data['expires_in'],
+				'access_token'   => $account_data['access_token'],
+				'refresh_token'  => $account_data['refresh_token'],
+			),
+			'available_accounts' => $accounts,
+		);
+
+		return true;
+	}
+
+	/**
+	 * Method to refresh access token.
+	 *
+	 * @since  8.5.9
+	 * @access private
+	 *
+	 * @return array
+	 */
+	private function gmb_refresh_access_token() {
```

Comment:
maybe this should be part of the auth service architecture we had discussed. Anything using ROP_AUTH_APP_URL should become part of this.